nixops/util.py

```diff
@@ -59,6 +72,38 @@ def __str__(self) -> str:
         return "{0} (exit code {1})".format(self.message, self.exitcode)
 
 
+class ImmutableMapping(Mapping):
```

Comment:
looks good but let's get a test on this. I don't quite understand it all, and I think some tests will help us make sure we understand / don't break what is going on.